torch/csrc/jit/script/module.cpp

```diff
@@ -237,6 +237,29 @@ Module Module::clone_impl(
   return r;
 }
 
+Module Module::clone_instance() const {
+  Module r(_ivalue()->compilation_unit(), type());
+
+  // Copy slots. If a slot is a module - recursively clone it.
+  size_t N = type()->numAttributes();
+  for (size_t i = 0; i < N; ++i) {
+    IValue s = _ivalue()->getSlot(i);
+    if (type()->getAttribute(i)->is_module()) {
+      const Module& orig = Module(s.toObject());
+      Module cloned = orig.clone_instance();
+      r.register_module(type()->getAttributeName(i), cloned);
```

Comment:
It would be better to not use `register_*` API here and explicitly use `setattr` on the object itself. The `clone_instance` function should not be able to mutate the type, but `register_`* is a mutating API. I realize that it doesn't change it in this particular case, but it still would be cleaner not to use it at all. Think about the type as having `const` modifier in this function and `register_` methods are not marked `const` so we're getting a compilation error here :)